### ISSUE TITLE
Grid peak shaving value problem

### ISSUE THREAD

**Issue by terzo33** [davidrapan/ha-solarman#262]
Hi, when i change grid peak shaving value, give me an error

![Screenshot_20241114_095100_Home Assistant.jpg](https://github.com/user-attachments/assets/03940a7c-f864-4879-8795-13df9f830af0)

I'm using logger LAN adapter

**Comment by davidrapan** [comment]
Hi, can you try the change with debug logging enabled? Thanks.

**Comment by terzo33** [comment]
> Hi, can you try the change with debug logging enabled? Thanks.

yes, I'm starting to record now

**Comment by terzo33** [comment]
> Hi, can you try the change with debug logging enabled? Thanks.

@davidrapan see log... thanks

![home-assistant.txt](https://github.com/user-attachments/files/17756880/home-assistant.txt)

those spikes that go to 0 are due to the integration stopping working for 5 seconds and then working again

**Comment by terzo33** [comment]
@davidrapan 12h log

![home-assistant_2.txt](https://github.com/user-attachments/files/17771278/home-assistant_2.txt)

**Comment by davidrapan** [comment]
Hi @githubDante, could you take a look at it?

**Comment by davidrapan** [comment]
@terzo33 can you please try to set the value using built-in ACTIONS in Developer tools. Use Modbus function code 0x06.

**Comment by githubDante** [comment]
@terzo33  can you confirm that the inverter is working according to the selected value despite the error which you see at the bottom of the screen

**Comment by terzo33** [comment]
> @terzo33  can you confirm that the inverter is working according to the selected value despite the error which you see at the bottom of the screen 

Yes confirm... works but I see error a the bottom of the screen. Need to fix it

**Comment by davidrapan** [comment]
So it does change the value in your inverter? Then ignore the error for now and I'll try to do something about it later. 😉 

Thanks @githubDante!

**Comment by terzo33** [comment]
@davidrapan when I set the value, I get the error, however after 10/20 seconds the value is set correctly on the inverter. Before with the wifi logger, everything was immediate. Also every now, the solarman integration stops working for a few seconds using the LAN logger.

**Comment by davidrapan** [comment]
> when I set the value, I get the error, however after 10/20 seconds the value is set correctly on the inverter.

The writing into inverter works correctly. The error you see is from parsing of the response from inverter. Nothing serious.

> Also every now, the solarman integration stops working for a few seconds using the LAN logger.

This is worse... Maybe slight adjustments to the timings of the requesting flow could improve it so you don't see disconnects.

**Comment by terzo33** [comment]
@davidrapan I hope you can fix it

**Comment by davidrapan** [comment]
In const.py you can try to change rows:
```
TIMINGS_UPDATE_TIMEOUT = TIMINGS_INTERVAL * 6
TIMINGS_SOCKET_TIMEOUT = TIMINGS_INTERVAL * 4 - 1
```

To:
```
TIMINGS_UPDATE_TIMEOUT = TIMINGS_INTERVAL * 8
TIMINGS_SOCKET_TIMEOUT = TIMINGS_INTERVAL * 6 - 1
```

**Comment by terzo33** [comment]
@davidrapan done

**Comment by davidrapan** [comment]
Don't forget to restart HA to apply changes and let me know if it had any effect on the disconnects.

**Comment by terzo33** [comment]
> Don't forget to restart HA to apply changes and let me know if it had any effect on the disconnects.

@davidrapan done but it's the same

![Screenshot_20241116_112751_Home Assistant.jpg](https://github.com/user-attachments/assets/b82d5ed8-94f5-40f0-94e0-579217fefac7)

**Comment by terzo33** [comment]
> Well in that case I don't think we can do anything about it in integration... but it's really strange that it goes to 0W...

@davidrapan I checked better and there is no connection problem but simply the parameters all go to 0, then after a few minutes they return to the correct values. there will surely be a bug

![Screenshot_20241118_102348_Samsung Internet.jpg](https://github.com/user-attachments/assets/e843b195-4408-41e3-b516-b0b4df0d46d0)

**Comment by davidrapan** [comment]
And all the sensors goes to 0 too?

**Comment by terzo33** [comment]
> And all the sensors goes to 0 too?

@davidrapan only program time, program power, program soc and program voltage

**Comment by davidrapan** [comment]
So the sensors shows what? Value or Unavailable?

**Comment by terzo33** [comment]
> So the sensors shows what? Value or Unavailable?

@davidrapan show 0 values

![Screenshot_20241118_102348_Samsung Internet.jpg](https://github.com/user-attachments/assets/a165d416-66b2-4c85-8e67-283b68207fbc)

**Comment by davidrapan** [comment]
Those are not sensors.

**Comment by terzo33** [comment]
> Those are not sensors.

Sorry, all sensors works fine, show alwasys correct value

**Comment by davidrapan** [comment]
I have honestly no idea why is it happening. It would make sense if also the sensor values did become Unavailable during periods with zeroed configurables...

You switched from Wifi Logger to Ethernet Logger?

**Comment by terzo33** [comment]
Hi @davidrapan , after upgrade solarman integration I see always this error

![Screenshot_20241205_090907_GitHub.jpg](https://github.com/user-attachments/assets/7b308ee0-e827-4427-9cf5-6db17e127335)

**Comment by davidrapan** [comment]
What does the log say?

**Comment by terzo33** [comment]
> What does the log say?
![home-assistant_3.txt](https://github.com/user-attachments/files/18020080/home-assistant_3.txt)

I see some error for "Buffer of 4 bytes error" and then I see on dashboard all editabile value go to 0 already :(

![Screenshot_20241205_094547_Home Assistant.jpg](https://github.com/user-attachments/assets/8263fe8a-b12a-4c55-b0c6-b8fe2afa8afa)

**Comment by davidrapan** [comment]
No I mean i want to see the debug log during changing the value. 😉

**Comment by terzo33** [comment]
> No I mean i want to see the debug log during changing the value. 😉

I sent you

![home-assistant_3.txt](https://github.com/user-attachments/files/18020402/home-assistant_3.txt)

**Comment by davidrapan** [comment]
> I sent you

Ah my bad, did not noticed that.

**Comment by davidrapan** [comment]
Do you have this issue with all of the number configurables?

**Comment by terzo33** [comment]
> Do you have this issue with all of the number configurables?

@davidrapan yes :(

**Comment by davidrapan** [comment]
I'm not sure but it could be specific to Ethernet Loggers... :-/ Cause I'm not able to replicate it on my device.

Are you familiar with python? Cause you proly gonna have to debug this on your installation.

**Comment by githubDante** [comment]
I think this is not needed at all. The request is correct, but the response from the datalogger after write is truncated. 

Here an example:

```
2024-12-05 09:47:40.406 DEBUG (MainThread) [custom_components.solarman.api] [2110011481] SENT: 3e 21 00 00 00 09 01 10 01 25 00 01 02 0a f0
2024-12-05 09:47:40.670 DEBUG (MainThread) [custom_components.solarman.api] [2110011481] RECD: 3e 21 00 00 00 04 01 10 01 25
```

```python
>>> pdu = bytes.fromhex('10 01 25 00 01 02 0a f0')
>>> func = create_function_from_request_pdu(pdu)
>>> func
<umodbus.functions.WriteMultipleRegisters object at 0x7349e5708e80>
>>> func.starting_address
293
>>> func.values
[2800]
>>> func.create_response_pdu().hex(' ')
'10 01 25 00 01'
```
As you can see two bytes are missing from the end.

**Comment by terzo33** [comment]
yes indeed solarman gives the error but then it sets the value correctly

**Comment by davidrapan** [comment]
Ah I did not notice that! I immediately assumed the error was in my code. 😆 Thanks!

**Comment by terzo33** [comment]
> Ah I did not notice that! I immediately assumed the error was in my code. 😆 Thanks!

Awesome. Could you fix for next release?

**Comment by davidrapan** [comment]
I totally forgot about and thought it's new. Guess I'm overworked. 😆

**Comment by davidrapan** [comment]
Do we have debug log with that happening?

**Comment by githubDante** [comment]
> So anyway, I was right and it's specific to Ethernet Loggers.

I wouldn't count on that. Can be a firmware bug :smiley: 

@terzo33 sorry but no, I don't use HA at all.

**Comment by davidrapan** [comment]
> I wouldn't count on that. Can be a firmware bug 😃

Sure, what I mean is we are not talking solarman protocol (it's in a different league all together and adds bytes instead, haha) here.

**Comment by davidrapan** [comment]
So this is how the reply should look like right?
```
8e 39 00 00 00 06 01 10 01 25 00 01
```

**Comment by githubDante** [comment]
Yes, that's the correct response. Here a small patch which should fix this

```patch
diff --git a/custom_components/solarman/api.py b/custom_components/solarman/api.py
index d59adaf..1ad82f5 100644
--- a/custom_components/solarman/api.py
+++ b/custom_components/solarman/api.py
@@ -101,7 +101,11 @@ class PySolarmanV5AsyncWrapper(PySolarmanV5Async):
     async def write_multiple_holding_registers(self, register_addr, values):
         if not self._passthrough:
             return await super().write_multiple_holding_registers(register_addr, values)
-        return await self._tcp_parse_response_adu(write_multiple_registers(self.mb_slave_id, register_addr, values))
+        try:
+            r = await self._tcp_parse_response_adu(write_multiple_registers(self.mb_slave_id, register_addr, values))
+        except struct.error:
+            r = len(values)
+        return r
 
 class Inverter(PySolarmanV5AsyncWrapper):
     def __init__(self, address, serial, port, mb_slave_id):
```
Don't know why I can't attach it...

**Comment by davidrapan** [comment]
I made [fix: Workaround for modbus tcp responses w/o quantity](https://github.com/davidrapan/ha-solarman/commit/e68423ebdc2c822ab20325dd607a3f237127fbdd), try it and if it fails I'll got w/ @githubDante's suggestion. 😉

Anyway, really thanks, Modbus would be sometimes complete madness for me without your help. 😆

**Comment by davidrapan** [comment]
Hi @terzo33, please try latest version and let me know if that bug of printing errors is gone. 😉

**Comment by terzo33** [comment]
> Hi @terzo33, please try latest version and let me know if that bug of printing errors is gone. 😉

@davidrapan yes works 😍😍 but in grid power i see something wrong 😂😂

![Screenshot_20241206_211939_Home Assistant.jpg](https://github.com/user-attachments/assets/b1da1b17-771b-4939-b666-a6ee48bf0cc7)

**Comment by davidrapan** [comment]
Please, can you tell me your exact device and model?

**Comment by terzo33** [comment]
> Please, can you tell me your exact device and model?

Deye Hybrid 6kW

**Comment by davidrapan** [comment]
I need the model number exactly.

**Comment by terzo33** [comment]
> I need the model number exactly.
> 
> Edit1: It's something like SUN-12K-SG04LP3 (this is mine)
> 
> Edit2: And can you show me the same screshoots like in https://github.com/davidrapan/ha-solarman/issues/281#issuecomment-2520017035 togerther with screeshots of whole sensor list?

SUN-6K-SG03LP1-EU

![Screenshot_20241206_213437_Home Assistant.jpg](https://github.com/user-attachments/assets/ae3665bf-134e-4cc2-b547-ff18aee30b28)

**Comment by davidrapan** [comment]
And screenshot of Grid and Load sensors? (Voltage, current, power)

**Comment by davidrapan** [comment]
Mhmmm, gonna have to find out how to detect which device needs scaling * 10 for Grid and Load power and which do not...

Thanks!

**Comment by terzo33** [comment]
> Mhmmm, gonna have to find out how to detect which device needs scaling * 10 for Grid and Load power and which do not...
> 
> Thanks!

I'm glad to help. If can help you. before the update the value was correct

**Comment by davidrapan** [comment]
I know, it was requested change because some devices do need it multiplied by 10.
